src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -150,65 +150,65 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
 	{
 		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
 		bool singleReplicatedTable = true;
+		bool referencedIsDistributed = false;
+		bool referencedIsReferenceTable = false;
 
 		if (constraintForm->contype != CONSTRAINT_FOREIGN)
 		{
 			heapTuple = systable_getnext(scanDescriptor);
 			continue;
 		}
 
+		referencedTableId = constraintForm->confrelid;
+		selfReferencingTable = (referencingTableId == referencedTableId);
+
+		referencedIsDistributed = IsDistributedTable(referencedTableId);
+		if (referencedIsDistributed)
+		{
+			referencedIsReferenceTable =
+				(PartitionMethod(referencedTableId) == DISTRIBUTE_BY_NONE);
+		}
+		else if (selfReferencingTable)
+		{
+			/*
+			 * If the table is just being distributed, IsDistributedTable() returns
+			 * false, so we need a special check for self-referencing tables.
+			 */
+			referencedIsReferenceTable = (distributionMethod == DISTRIBUTE_BY_NONE);
```

Comment:
could we fill in the distributionMethod at the start if the table already exists to avoid the extra call to PartitionMethod above?

Reply:
> could we fill in the distributionMethod at the start if the table already exists to avoid the extra call to PartitionMethod above?

`PartitionMethod` was called once per referenced relation. Since this function was very confusing, I did a total overhaul of it after your comment. Trying to make things a bit more obvious. The logic is divided into three functions:

1. ErrorIfUnsupportedForeignConstraintExists -> Does the high level looping and information preparation.
2. ErrorIfUnsupportedForeignConstraint -> Our main logic
3. ForeignConstraintFindDistKeys -> A low-level detail which took too much space, so I thought it can be refactored out.

The logic is same as before, but in 3 separate functions.